galsim/phase_psf.py

```diff
@@ -0,0 +1,1017 @@
+# Copyright (c) 2012-2015 by the GalSim developers team on GitHub
+# https://github.com/GalSim-developers
+#
+# This file is part of GalSim: The modular galaxy image simulation toolkit.
+# https://github.com/GalSim-developers/GalSim
+#
+# GalSim is free software: redistribution and use in source and binary forms,
+# with or without modification, are permitted provided that the following
+# conditions are met:
+#
+# 1. Redistributions of source code must retain the above copyright notice, this
+#    list of conditions, and the disclaimer given in the accompanying LICENSE
+#    file.
+# 2. Redistributions in binary form must reproduce the above copyright notice,
+#    this list of conditions, and the disclaimer given in the documentation
+#    and/or other materials provided with the distribution.
+#
+"""@file phase_psf.py
+Utilities for creating PSFs from phase screens.  Essentially evaluates the Fourier optics
+diffraction equation:
+
+PSF(x, y) = int( |FT(aperture(u, v) * exp(i * phase(u, v, x, y, t)))|^2, dt)
+
+where x, y are focal plane coordinates and u, v are pupil plane coordinates
+
+The main classes of note are:
+
+FrozenAtmosphericScreen
+  Class implementing phase(u, v, x, y, t) for frozen flow von Karman type turbulence
+
+ARAtmosphericScreen
+  Class implementing phase(u, v, x, y, t) for "boiling" von Karman type turbulence
+
+PhaseScreenList
+  Python sequence type to hold multiple phase screens, for instance to simulate turbulence at
+  different altitudes.  A key method is getPSF(), which will take the list of phase screens,
+  add them together linearly (Fraunhofer approximation), and evaluate the above diffraction
+  equation to yield a PhaseScreenPSF object.
+
+PhaseScreenPSF
+  A GSObject holding the evaluated PSF.
+
+Atmosphere
+  Convenience function to quickly assemble multiple AtmosphericScreens into a PhaseScreenList.
+"""
+
+import copy
+import numpy as np
+import galsim
+import utilities
+from galsim import GSObject
+
+
+def listify(arg):
+    """Turn argument into a list if not already iterable."""
+    return [arg] if not hasattr(arg, '__iter__') else arg
+
+
+def broadcast(arg, n):
+    """Turn length-1 iterable into length-n list."""
+    return [arg[0]]*n if len(arg) == 1 else arg
+
+
+def generate_pupil(nu, pupil_size, diam=None, obscuration=None):
+    """ Generate a pupil transmission array (0's and 1's) for a circular aperture and potentially a
+    central circular obscuration.
+
+    @param nu          Number of pixels across pupil array.
+    @param pupil_size  Physical size of pupil array in meters.
+    @param diam        Diameter of aperture in meters.
+    @param obscuration Fractional linear obscuration.
+    @returns array of 0's and 1's indicating pupil transmission function.
+    """
```

Comment:
There is already a function that does much the same thing (although also with struts) in optics.  Can we maybe find a way to use that function here, rather than rewrite this functionality?  Feel free to refactor some of the functions in optics.py if needed, but let's try to avoid code duplication if possible.